### ISSUE TITLE
Unit Test Assert does not allow to check method calls for idempotence

### ISSUE THREAD

**Issue by PeregrinTooc** [SAP/code-pal-for-abap#460]
> 💡 Code pal for ABAP follows the [Clean ABAP](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md). If the issue relates to coding style, please submit it [here](https://github.com/SAP/styleguides/issues). 

**Check Name**
[unit test assert](https://github.com/SAP/code-pal-for-abap/blob/master/docs/checks/unit_test_assert.md)

**Actual Behavior**
If the same method is called twice in actual and expected, e.g. to test that a factory method returns a singleton, 'illegal assert' is raised

**Expected Behavior**
For method calls, dispense with the rule, since idempotence is a meaningful test.

**Comment by lucasborin** [comment]
Unfortunately, Code Pal performs a static check (string-based), and it cannot detect idempotent calls.
What we could implement is exempting if the act&exp ends with `get_instance( )`. How about it?

**Comment by pokrakam** [comment]
I think that's too specific, you can also have
```ABAP
cl_abap_unit_assert=>assert_differt( 
  act = get_next( )
  exp = get_next( ) ).
```
But I also think this is a narrow use case that is easy enough to work around by assigning one of the values to a variable.

**Comment by pokrakam** [comment]
Actually I just spotted that we're only talking about method calls. In this case it should be sufficient to exclude anything that has a "(" in it, no?

**Comment by lucasborin** [comment]
> Actually I just spotted that we're only talking about method calls. In this case it should be sufficient to exclude anything that has a "(" in it, no?

I am with you guys, it makes sense to disable this rule on these method calls due to this technical limitation. I am unsure about the `(` because you could compare a slice of a string in the `act` vs `exp`, like: 

```abap
DATA(memory) = '010101110101'.

cl_abap_unit_assert=>assert_differs( act = memory+2(1)
                                     exp = memory+3(1) ).
```

**Comment by lucasborin** [comment]
I was wondering to check the tokens `=>` and `->` as usually you will call a `cut` method. However, local methods won't be supported by this rule. 

```abap
cl_abap_unit_assert=>assert_equals( act = my_class=>get_instance( )
                                    exp = my_class=>get_instance( ) ).
```

**Comment by pokrakam** [comment]
If you're treating it as a string then matching `^[a-z0-9_]+\(` should avoid the substring variation.

**Comment by PeregrinTooc** [comment]
Why not just extend it not to trigger if it ends with ‚( ).‘ so all method
calls are exempt?

Lucas Borin ***@***.***> schrieb am Mi. 15. Sept. 2021 um
16:03:

**Comment by PeregrinTooc** [comment]
Oh wait, too late for me to really think straight 🙈 parameters might be
there 😂

Philipp Rembold ***@***.***> schrieb am Mi. 15. Sept. 2021
um 22:36:

**Comment by pokrakam** [comment]
Yes, on second thoughts it's a little more complex. The expression `[a-z_][a-z0-9_]*\(` should do it for more complex expressions.
It ignores `foo && bar+2(1)` but matches `cls=>foo && obj->bar( ) && moo+2(1)`

**Comment by lucasborin** [comment]
@PeregrinTooc fix available under branch `1.16.0`. Feel free to test it. 😃
